dao-voting/src/main.sw

```diff
@@ -0,0 +1,295 @@
+        match sender {
+            Identity::Address(address) => transfer_to_output(amount, storage.gov_token, address), Identity::ContractId(address) => force_transfer_to_contract(amount, storage.gov_token, address), 
+        }
+
+        true
+    }
+
+    /// Vote on a given proposal
+    ///
+    /// # Parameters
+    ///
+    /// proposal_id - proposal to vote on
+    /// vote_amount - amount of votes to use on proposal
+    /// is_yes_vote - determines if you vote yes or no on the proposal
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The proposal id is out of range
+    /// - The vote amount is 0
+    /// - The vote amount is greater than the users deposited balance
+    /// - The given proposal is expired
+    #[storage(read, write)]fn vote(proposal_id: u64, vote_amount: u64, is_yes_vote: bool) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+        require(vote_amount > 0, Error::VoteAmountCannotBeZero);
+
+        let sender: Identity = sender_identity();
+        let sender_balance = storage.balances.get(sender);
+
+        require(sender_balance >= vote_amount, Error::NotEnoughAssets);
+
+        let mut proposal = storage.proposals.get(proposal_id);
+        require(proposal.end_height >= height(), Error::ProposalExpired);
+
+        if (is_yes_vote) {
+            proposal.yes_votes = proposal.yes_votes + vote_amount;
+        } else {
+            proposal.no_votes = proposal.no_votes + vote_amount;
+        };
+
+        storage.balances.insert(sender, sender_balance - vote_amount);
+
+        let votes = storage.votes.get((sender, proposal_id));
+        storage.votes.insert((sender, proposal_id), votes + vote_amount);
+
+        storage.proposals.insert(proposal_id, proposal);
+
+        true
+    }
+
+    /// Execute a given proposal
+    ///
+    /// # Parameters
+    ///
+    /// proposal_id - proposal to execute
+    ///
+    /// # Panics
+    ///
+    /// This function will panic when:
+    /// - The construct has not been called to initialize
+    /// - The proposal id is out of range
+    /// - The proposal has not expired
+    /// - The proposal has not met the necessary approval percentage
+    #[storage(read, write)]fn execute(proposal_id: u64) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+
+        let proposal = storage.proposals.get(proposal_id);
+        require(proposal.end_height < height(), Error::ProposalActive);
+
+        // TODO figure out how to prevent approval percentage from overflowing
+        // When close to the u64 max
+        // https://github.com/FuelLabs/sway-applications/issues/106
+        let approval_percentage = proposal.yes_votes * 100 / (proposal.yes_votes + proposal.no_votes);
```

Comment:
yes and no votes can both be 0 here. should this be handled?

Reply:
Wait, should you be adding these or subtracting?

Reply:
If both yes and no votes are 0 we should probably handle it here.  Currently if both yes and no votes are both 0 this line will revert with an arithmetic error I'm guessing.  However,  if both yes and no votes are 0 there is no possible way for the approval percentage to be met (it is an error to add a proposal with 0 approval_percentage), which would revert the function anyways.  That being said do you think it is still worth it to handle this edge case?

Regarding your second comment, I believe we should be adding these.  The approval_percentage for a proposal is the number of yes votes out of the total number of votes.  It might be a good idea for me to document this line more clearly.    If this line was in a regular backend application I would use floats and just do yes_votes / (total_votes) to get the approval percentage.  Let me know if you need more clarification.

Reply:
I guess reverting automatically when they are both 0 is fine. Whenever I see a division by 0 in a formula I wonder about the black holes that we are spawning somewhere and whether the universe will implode.